pkg/apiserver/middlewares/v1/api_key.go

```diff
@@ -49,34 +53,143 @@ func HashSHA512(str string) string {
+			//This is likely not the proper way, but isNotFound does not seem to work
+			if err != nil && strings.Contains(err.Error(), "bouncer not found") {
+				//Because we have a valid cert, automatically create the bouncer in the database if it does not exist
+				//Set a random API key, but it will never be used
+				//To fix
+				apiKey, err := GenerateAPIKey(64)
+				if err != nil {
+					log.WithFields(log.Fields{
+						"ip": c.ClientIP(),
+						"cn": extractedCN,
+					}).Errorf("error generating mock api key: %s", err)
+					c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+					c.Abort()
+					return
+				}
+				log.WithFields(log.Fields{
+					"ip": c.ClientIP(),
+					"cn": extractedCN,
+				}).Infof("Creating bouncer %s", bouncerName)
+				err = a.DbClient.CreateBouncer(bouncerName, c.ClientIP(), HashSHA512(apiKey), types.TlsAuthType)
+				if err != nil {
+					log.WithFields(log.Fields{
+						"ip": c.ClientIP(),
+						"cn": extractedCN,
+					}).Errorf("creating bouncer db entry : %s", err)
+					c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+					c.Abort()
+					return
+				}
+				bouncer, err = a.DbClient.SelectBouncerByName(bouncerName)
+				if err != nil {
+					log.WithFields(log.Fields{
+						"ip": c.ClientIP(),
+						"cn": extractedCN,
+					}).Errorf("auth api key error: %s", err)
+					c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+					c.Abort()
+					return
+				}
+			} else if err != nil {
+				//error while selecting bouncer
+				log.WithFields(log.Fields{
+					"ip": c.ClientIP(),
+					"cn": extractedCN,
+				}).Errorf("while selecting bouncers: %s", err)
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			} else {
+				//bouncer was found in DB
+				if bouncer.AuthType != types.TlsAuthType {
+					log.WithFields(log.Fields{
+						"ip": c.ClientIP(),
+						"cn": extractedCN,
+					}).Errorf("bouncer isn't allowed to auth by TLS")
+					c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+					c.Abort()
+					return
+				}
+			}
+		} else {
+			//API Key Authentication
+			val, ok := c.Request.Header[APIKeyHeader]
+			if !ok {
+				log.WithFields(log.Fields{
+					"ip": c.ClientIP(),
+				}).Errorf("API key not found")
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			}
+			hashStr := HashSHA512(val[0])
+			bouncer, err = a.DbClient.SelectBouncer(hashStr)
+			if err != nil {
+				log.WithFields(log.Fields{
+					"ip": c.ClientIP(),
+				}).Errorf("while fetching bouncer info: %s", err)
+				c.JSON(http.StatusForbidden, gin.H{"message": "access forbidden"})
+				c.Abort()
+				return
+			}
+			if bouncer.AuthType != types.ApiKeyAuthType {
+				log.WithFields(log.Fields{
+					"ip": c.ClientIP(),
+				}).Errorf("bouncer %s attempted to login using an API key but it is configured to auth with TLS cert", bouncer.Name)
```

Comment:
maybe we can print the `bouncer.AuthType` instead of "TLS cert" in case something is going wrong ?